File: src/javascript/algorithms/strings/minWindowSubstring.js
// Sliding window: TC: O(m) + O(n) SC: O(m) + O(n)

function minWindowSubstring(windowStr, subStr) {
    let windowStrCount = new Map();
    let subStrCount = new Map();

    let minLength = Number.MAX_VALUE;
    let leftIndex = rightIndex = -1;

    for(const ch of subStr) {
        subStrCount.set(ch, (subStrCount.get(ch) || 0) + 1);
    }

    let having = 0; required = subStrCount.size;
    let left = right = 0;

    for(right = 0; right < windowStr.length; right++) {
        let rightChar = windowStr.charAt(right);
        windowStrCount.set(rightChar, (windowStrCount.get(right) || 0) +1);
        if(subStrCount.has(rightChar) && subStrCount.get(rightChar) === windowStrCount.get(rightChar)) {
            having++;
        }

        while(required === having) {
            if(minLength > right-left+1) {
                minLength = right-left+1;
                leftIndex = left;
                rightIndex = right;
            }

            let leftChar = windowStr.charAt(left);
            windowStrCount.set(leftChar, windowStrCount.get(leftChar)-1);
            if(subStrCount.has(leftChar) && subStrCount.get(leftChar) > windowStrCount.get(leftChar)) {
                having--;
            }
            left++;
        }
    }
    if(leftIndex === -1 || rightIndex === -1) {
        return "";
    } else {
        return windowStr.substring(leftIndex, rightIndex+1);
    }

}

let s ="ADOBECODEBANC";
let t= "ABC";
console.log(minWindowSubstring(s, t));

File: src/java1/algorithms/strings/minWindowSubstring/MinWindowSubstring.java
package java1.algorithms.strings.minWindowSubstring;
// Sliding window: TC: O(m) + O(n), SC: O(m) + O(n)

import java.util.HashMap;
import java.util.Map;

public class MinWindowSubstring {

    private static String minWindowSubstring(String windowStr, String subStr) {

        if (windowStr.isEmpty() || subStr.isEmpty() || windowStr.length() < subStr.length()) return "";

        Map<Character, Integer> subStrCount = new HashMap<>();
        Map<Character, Integer> windowStrCount = new HashMap<>();

        //Step1: Build frequency map of target substring
        for(char ch: subStr.toCharArray()) {
            subStrCount.put(ch, subStrCount.getOrDefault(ch, 0) + 1);
        }

        //Step2: Initialize sliding window
        int minLength = Integer.MAX_VALUE;
        int[] bestWindow = new int[]{-1, -1};

        int having = 0, required = subStrCount.size();
        int left = 0;

        //Step3: Expand window and update counts
        for(int right = 0; right < windowStr.length(); right++) {
            char rightChar = windowStr.charAt(right);
            
            if(subStrCount.containsKey(rightChar)) {
                windowStrCount.put(rightChar, windowStrCount.getOrDefault(rightChar, 0) + 1);
                if(subStrCount.get(rightChar) == windowStrCount.get(rightChar)) {
                    having++;
                } 
            }

            //Step4: Shrink window to find minimized substring
            while(required == having) {
                if(minLength > right-left+1) {
                    minLength = right-left+1;
                    bestWindow[0] = left;
                    bestWindow[1] = right;
                }
                
                char leftChar = windowStr.charAt(left);
                if(subStrCount.containsKey(leftChar)) {
                    windowStrCount.put(leftChar, windowStrCount.get(leftChar)-1);
                    if(windowStrCount.get(leftChar) < subStrCount.get(leftChar)) {
                        having--;
                    }
                }
                left++;
            }
        }

        return minLength == Integer.MAX_VALUE ? "" : windowStr.substring(bestWindow[0], bestWindow[1]+1);
    }

    public static void main(String[] args) {
        String s1 ="ADOBECODEBANC", t1 = "ABC";
        System.out.println(minWindowSubstring(s1, t1));

        String s2 ="A", t2 = "A";
        System.out.println(minWindowSubstring(s2, t2));

        String s3 ="a", t3 = "aa";
        System.out.println(minWindowSubstring(s3, t3));
    }
}
